lib/ssl/src/tls_connection_1_3.erl

```diff
@@ -257,41 +257,72 @@ user_hello(_, _, _) ->
+start(internal, #client_hello{extensions = #{client_hello_versions :=
+                                                 #client_hello_versions{versions = ClientVersions}} = Extensions} = Hello,
+                              #state{ssl_options = #{handshake := hello},
              start_or_recv_from = From,
              handshake_env = HsEnv} = State) ->
-    {next_state, user_hello,
-     State#state{start_or_recv_from = undefined,
-                 handshake_env = HsEnv#handshake_env{
-                                   hello = Hello}}, [{reply, From, {ok, Extensions}}]};
-start(internal, #client_hello{} = Hello, State0) ->
-    case tls_handshake_1_3:do_start(Hello, State0) of
-        #alert{} = Alert ->
-            ssl_gen_statem:handle_own_alert(Alert, start, State0);
-        {State, start} ->
-            {next_state, start, State, []};
-        {State, negotiated} ->
-            {next_state, negotiated, State, [{next_event, internal, {start_handshake, undefined}}]};
-        {State, negotiated, PSK} ->  %% Session Resumption with PSK
-            {next_state, negotiated, State, [{next_event, internal, {start_handshake, PSK}}]}
+    case tls_record:is_acceptable_version({3,4}, ClientVersions) of
+        true ->
+            {next_state, user_hello,
+             State#state{start_or_recv_from = undefined,
+                         handshake_env = HsEnv#handshake_env{
+                                           hello = Hello}}, [{reply, From, {ok, Extensions}}]};
+        false ->
+             ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?PROTOCOL_VERSION), start, State)
     end;
-start(internal, #server_hello{extensions = Extensions} = ServerHello, 
-      #state{ssl_options = #{handshake := hello},
+start(internal, #client_hello{extensions = #{client_hello_versions :=
+                                                 #client_hello_versions{versions = ClientVersions}
+                                            }} = Hello, State0) ->
+    case tls_record:is_acceptable_version({3,4}, ClientVersions) of
+        true ->
+            case tls_handshake_1_3:do_start(Hello, State0) of
+                #alert{} = Alert ->
+                    ssl_gen_statem:handle_own_alert(Alert, start, State0);
+                {State, start} ->
+                    {next_state, start, State, []};
+                {State, negotiated} ->
+                    {next_state, negotiated, State, [{next_event, internal, {start_handshake, undefined}}]};
+                {State, negotiated, PSK} ->  %% Session Resumption with PSK
+                    {next_state, negotiated, State, [{next_event, internal, {start_handshake, PSK}}]}
+            end;
+        false ->
+            ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?PROTOCOL_VERSION), start, State0)
+        end;
+start(internal, #client_hello{}, State0) ->
+    ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?PROTOCOL_VERSION), ?FUNCTION_NAME, State0);
```

Comment:
I would like a comment for each clause matching `#client_hello`explaining what it means and when it can occur.
As I understand it the clauses handle the following cases:
1. The user has requested to handle part of the handshake indicated by existence of the `{handshake, hello}` option. We  return to the user which can continue the handshake
2. The client has sent which versions it wants to negotiate and we will check it that is acceptable
3. ??? When can this happen (a `#client_hello{}`without the extension `client_hello_versions`

I suggest that we merge the first 2 clauses together and that the test regarding the `handshake` option is done only if `tls_record:is_acceptable_version({3,4}, ClientVersions)` returns `true`.

Reply:
Sure I will add some comments. I actually think there are some more things that we might need to think about for this issue so I will work on a new version.

Reply:
I made a new version now, that will avoid double-checking the version when {handshake, hello} is used. I took the opportunity to get rid of a legacy mechanism and use gen_statem to postpone the handling of the client/server-hello until we receive the continuation with possible new options. I think this can also warrant some work on version handling in TLS-1.2 with considerations of adoption of some TLS-1.3 mechanisms mentioned in the RFC that preferably should affect also TLS-1.2 which I am not certain that we have adopted. This I think should be handled as a separate issue I will make a ticket.